### DISCUSSION TITLE
How to test MsgEffect?

### DISCUSSION THREAD

**Discussion by sukazavr** [marblejs/marble#353] [Q&A] [answered]
Hello!
What is the proper way to test MsgEffect since we only have HttpTestBed?

**Comment by JozefFlakus** [comment, marked as answer]
Hi @sukazavr!

`MessagingTestBed` was on its way but I postponed the implementation due to the ongoing process of introducing version 4.0.
So far the scope for the next major release doesn't include it but since it is a non breaking feature it can be added in the next 4.1 version. But directly answering your question...

`MsgEffect` is just a function. It takes `A` and returns `B`. You can write a simple utility function that packs an incoming `Event` into the stream and then feeds it in. Additionaly the `EffectContext` can be easily mocked. The only thing that is worth providing is the `Context` object that you can build using `contextFactory` function from `@marblejs/core` and passing it through `lookup`  function from the same module in order to conform to the `ContextProvider` / `ask` interface.

```typescript
import { contextFactory, lookup, createEffectContext } from '@marblejs/core';

const context = await contextFactory([
  // ... registered dependencies
]);

const ask = lookup(context);

// ...

const effectContextMock = createEffectContext({ ask, client: {} as any });

// ...

someEffect$(event$, effectContextMock).subscribe(
  // ...
);
```

Basically thats all building blocks that are needed. In practice the incoming `MessagingTestBed` will simplify the flow in order to not produce unnecessary boilerplate.
